lib/fs.ts

```diff
@@ -34,11 +34,46 @@ interface PendingUpdate {
 	pendingDataFetcher: (() => string | Buffer) | null;
 	pendingOptions: object | null;
 	throttleTime: number; // throttling until time (0 for no throttle)
-	throttleTimer: NodeJS.Timer | null;
+	throttleTimer: CancellableTimerPromise<void> | null;
+	finish: Promise<void>;
 }
 
 const pendingUpdates = new Map<string, PendingUpdate>();
 
+/**
+ * Givn an async function and ms, this will create a promise that resolves
+ * when the timer and async function ends
+ *
+ * It also has a `cancel` function that will clear the timer
+ * and resolve the promise
+ */
+class CancellableTimerPromise<T, U extends T | PromiseLike<T> | undefined = undefined> {
+	promise: Promise<T>;
+	private resolveVal: U | undefined;
+	// @ts-ignore - definitely assigned in the constructor
+	private timeout: NodeJS.Timeout;
+	// @ts-ignore - see above
+	private resolve: (value?: any) => void;
+	constructor(fn: () => Promise<T>, ms: number, resolveVal?: U) {
+		this.resolveVal = resolveVal;
```

Comment:
I would use

```suggestion
		this.resolveVal = resolveVal;
		this.timeout = null!;
		this.resolve = null!;
```
instead of your ignores.